# Data Mining
## Assignment 1 - Data Cleansing

---

## Aadam - CS1945
## Obaidullah - CS1947

---

## Tasks to do
1. Combine all the three datasets in a single source, you may use any DBMS. 
2. Once you have the data in the staging area inside your DBMS, perform data profiling for all the fields. By data profiling means the following statistics:
    - No. of unique values (for each column) 
    - No. of nulls (for each column) 
    - Invalid values (for each column) 
    - Total no. of courses 
    - Total no. of female students Vs male students 
    - Total no of people who has taken more than 5 courses in a semester. 
    - The relationship between total no. of unique student ID’s and total no. of students. 
    - Average no of people per semester of each campus. 
    - Average no of students in every batch of each campus.
3. After the profiling, you should have identified the anomalies and data cleansing issues. Here are some of the issues you need to address during your cleansing work.
    - Separate first and last names both for the student and father. Standardize the first and last names. (Hint: Find all the unique names in data and create a lookup table with two columns i.e. correct_name, variation etc. Use the lookup table to update the name fields with standardized names. Never hardcode names in your SQL.)
    - Introduce a new column i.e. city_name. It will involve extracting the city from the address and then the standardization of the city names. (Use the city names in the telephone directory as standard)
    - Bring the gender information into a consistent representation. Use ‘M’ and ‘F’ with data type CHAR(1).
    - Since the gender information is missing for the Peshawar campus, you can use the student names to figure out the gender. (Hint: Find all distinct male and female names and create a lookup table)
    - If you have been using VARCHAR or CHAR for date, port the date information into columns with the proper Date data type. The dates should be as per calendar dates.
    - Validate all the dates against the business rules e.g. the DOB should be smaller than the Reg. Date and Graduation Date should be greater than Reg. Date. The data might be having anomalies like exchanged DOB and Reg. Date by mistake etc. Also be careful with invalid dates i.e. 31st Feb. or 29th Feb. in a non-leap year.
    - For some campuses, the degree information is missing. Devise some technique to figure it out and update the rows with empty degree fields. Validate other businessrules for each field. One example is that marks should be in the range 0 to 100 inclusive.


In [323]:
import sqlite3

import pandas as pd

from pathlib import Path

In [324]:
data_dir = Path("Assignment 1 Data Set/")

In [800]:
con = sqlite3.connect('University.db')

df_std = pd.read_sql_query("SELECT * FROM Students", con)
df_reg = pd.read_sql_query("SELECT * FROM Course_Registrations", con)

con.close()

In [801]:
df_std.head()

SID                 Name                Father  \
0     KHR_BS_0       Hussain Ansary          Ubaid Ansary   
1     KHR_BS_1  Shk. Munir Hussaini  Viqar Hamid Hussaini   
2    KHR_BS_10         Hamna Ansary         Hameed Ansary   
3   KHR_BS_100        Jabbar Haqqie     Muneer Rai Haqqie   
4  KHR_BS_1000     Rana Hayyat Baig   Ghulam Mustafa Baig   

                   DoB Gender                    Reg Date Reg Status  \
0  1974-01-05 00:00:00      M  1994-08-13 00:00:00.000000          A   
1  1974-12-13 00:00:00      M  1994-08-13 00:00:00.000000          A   
2  1974-04-25 00:00:00      F  1994-08-14 00:00:00.000000          A   
3  1974-05-22 00:00:00      M  1994-08-18 00:00:00.000000          A   
4  1976-02-25 00:00:00      M  1995-08-28 00:00:00.000000          A   

  Degree Status                                            Address  \
0             C           h# 978 Street No.72  Defense Phase 1 KHR   
1             C                    H#  169 s#  0 Mutian wala Thata   
2             C                     h#  697 St. #  94 sea site KHR   
3             C                       H# 504 S No.4  Kumria Quetta   
4             C  H No.509 St. No.38  Kishwar heights university...   

  Qualification Degree Campus  
0       A-Level     BS    KHR  
1       A-Level     BS    KHR  
2       A-Level     BS    KHR  
3       A-Level     BS    KHR  
4          HSSC     BS    KHR

In [802]:
df_reg.head()

SID  Course  Score Semester                        Year Discipline  \
0  KHR_BS_0  CS-101     72     Fall  1994-01-01 00:00:00.000000         CS   
1  KHR_BS_0  CS-102     78     Fall  1994-01-01 00:00:00.000000         CS   
2  KHR_BS_0  CS-103     53     Fall  1994-01-01 00:00:00.000000         CS   
3  KHR_BS_0  CS-104     58     Fall  1994-01-01 00:00:00.000000         CS   
4  KHR_BS_0  CS-105     61   Spring  1995-01-01 00:00:00.000000         CS   

  Degree Campus  
0     BS    KHR  
1     BS    KHR  
2     BS    KHR  
3     BS    KHR  
4     BS    KHR

# Data Profiling

In [803]:
df_std.describe(include='all')

SID           Name            Father  \
count         17102          17102             17102   
unique        17102          14673             14270   
top     PEW_BS_1573  Kishwar Pasha  Mudassir Khattak   
freq              1              6                 5   

                               DoB Gender                    Reg Date  \
count                        17100  13398                       17102   
unique                        7380      2                         442   
top     1981-03-22 00:00:00.000000      M  2003-08-16 00:00:00.000000   
freq                             8   8935                          76   

       Reg Status Degree Status                                 Address  \
count       17102         17102                                   17102   
unique          2             2                                   15360   
top             A             C  house no.141 S# 16  Haider road Multan   
freq        17100         12500                                       3   

       Qualification Degree Campus  
count          17102  17102  17102  
unique            12      2      3  
top            F.Sc.     BS    KHR  
freq            2390  14302   8201

In [804]:
df_reg.describe(include='all')

SID  Course          Score Semester  \
count        406200  406200  406200.000000   406200   
unique        11700     120            NaN        2   
top     PEW_BS_1573  CS-103            NaN   Spring   
freq             48    8900            NaN   203100   
mean            NaN     NaN      74.460268      NaN   
std             NaN     NaN      14.415844      NaN   
min             NaN     NaN      50.000000      NaN   
25%             NaN     NaN      62.000000      NaN   
50%             NaN     NaN      74.000000      NaN   
75%             NaN     NaN      87.000000      NaN   
max             NaN     NaN      99.000000      NaN   

                              Year Discipline  Degree  Campus  
count                       406200     406200  406200  406200  
unique                          16          8       2       3  
top     1970-01-01 00:00:00.000002         TC      BS     LHR  
freq                         76200      94200  376800  190800  
mean                           NaN        NaN     NaN     NaN  
std                            NaN        NaN     NaN     NaN  
min                            NaN        NaN     NaN     NaN  
25%                            NaN        NaN     NaN     NaN  
50%                            NaN        NaN     NaN     NaN  
75%                            NaN        NaN     NaN     NaN  
max                            NaN        NaN     NaN     NaN

## No. of Unique Values in Students Table

In [805]:
df_std.nunique().to_frame().T

SID   Name  Father   DoB  Gender  Reg Date  Reg Status  Degree Status  \
0  17102  14673   14270  7380       2       442           2              2   

   Address  Qualification  Degree  Campus  
0    15360             12       2       3

## No. of Unique Values in Course Registration Table

In [806]:
df_reg.nunique().to_frame().T

SID  Course  Score  Semester  Year  Discipline  Degree  Campus
0  11700     120     50         2    16           8       2       3

## No. of Nulls in Students Table

In [797]:
pd.DataFrame(df_std.isnull().sum(axis = 0), columns=['Count']).T

SID  Name  Father  DoB  Gender  Reg Date  Reg Status  Degree Status  \
Count    0     0       0    2    3704         0           0              0   

       Address  Qualification  Degree  Campus  
Count        0              0       0       0


## No. of Nulls in Course Registration Table

In [796]:
pd.DataFrame(df_reg.isnull().sum(axis = 0), columns=['Count']).T

SID  Course  Score  Semester  Year  Discipline  Degree  Campus
Count    0       0      0         0     0           0       0       0

## Invalid Values

Let's try to see if the dates in DoB and Reg Date are valid.

In [770]:
dob = pd.to_datetime(df_std.DoB, errors="coerce")

In [468]:
df_std["Reg Date"] = pd.to_datetime(df_std["Reg Date"])

In [469]:
df_reg.Year = pd.to_datetime(df_reg.Year)

### Following are the invalid Date values in DOB

In [771]:
df_std[dob.isnull()]

SID                          Name  \
516    KHR_BS_1462                   Annan Satti   
593    KHR_BS_1531                 Hadiqa Abuzar   
1282   KHR_BS_2151                    Affan Baig   
2113    KHR_BS_290                   Qasim Rehan   
3631   KHR_BS_4266                  Shafaq Ahmed   
4167   KHR_BS_4749              Sumiyya Khanzada   
4319   KHR_BS_4886           Khushbakht Khanzada   
4320   KHR_BS_4887               Shabir Hussaini   
4933   KHR_BS_5438       Aga Abdus Sammi Khattak   
5623   KHR_BS_6059                Arsalan Chohan   
5852   KHR_BS_6265                  Amina Muslim   
7117   KHR_MS_1462       Mouhammad Ghani Paracha   
7194   KHR_MS_1531                    Arif Amjad   
7414    KHR_MS_290             Sonia Shaheen Haq   
10649  LHR_BS_3200  Choudhary Abdul Qadir Minhas   
11692   LHR_BS_414      Pervaiz Sikandar Sherazi   

                             Father        DoB Gender  \
516               Husain Alam Satti  29/Feb/78      M   
593                  Abuzar Zeeshan  29/Feb/77      F   
1282        Mohammad Sheharyar Baig  29/Feb/79      M   
2113                   Rehan Nabeel  29/Feb/75      M   
3631                     Ahmed Ajab  29/Feb/82      F   
4167                  Arif Khanzada  29/Feb/83      F   
4319            Sh. Sarwar Khanzada  29/Feb/83      F   
4320         Syyed Muazzum Hussaini  29/Feb/83      M   
4933     Abdul Qadeer Ikram Khattak  29/Feb/85      M   
5623              Rabbi Daud Chohan  29/Feb/86      M   
5852                 Muslim Shujaat  29/Feb/85      F   
7117           Rehmat Kabir Paracha  29/Feb/82      M   
7194                  Amjad Inayyat  29/Feb/81      M   
7414                    Shaheen Haq  29/Feb/78      F   
10649                  Viqar Minhas       None      M   
11692  Abdul Mannan Shabeer Sherazi       None      M   

                         Reg Date Reg Status Degree Status  \
516    1996-08-28 00:00:00.000000          A             C   
593    1996-08-22 00:00:00.000000          A             C   
1282   1997-08-16 00:00:00.000000          A             C   
2113   1994-08-29 00:00:00.000000          A             C   
3631   2001-08-21 00:00:00.000000          A             C   
4167   2001-08-03 00:00:00.000000          A             C   
4319   2002-08-16 00:00:00.000000          A             I   
4320   2002-08-19 00:00:00.000000          A             I   
4933   2003-08-20 00:00:00.000000          A             I   
5623   2004-08-18 00:00:00.000000          A             I   
5852   2004-08-24 00:00:00.000000          A             I   
7117   2004-08-28 00:00:00.000000          A             I   
7194   2004-08-22 00:00:00.000000          A             I   
7414   2001-08-29 00:00:00.000000          A             C   
10649  2002-08-20 00:00:00.000000          A             I   
11692  1995-08-28 00:00:00.000000          A             C   

                                                 Address Qualification Degree  \
516                  House No.483 St. No.37  Dehri Thata         F.Sc.     BS   
593                  h no. 292 St. No. 38 Koral Umer Kot          HSSC     BS   
1282        house no.568 St. No.52  Zeenat block Karachi         F.Sc.     BS   
2113       Ho. No. 193 St. No. 14 Chand mari Mirpur Khas       A level     BS   
3631           Ho. No.405 St. No.57  Doley shah Umer Kot       A level     BS   
4167    house no. 609 St. No. 98 Shah latif town Karachi         F.Sc.     BS   
4319             h no. 681 St. No. 4 Pwd Coloney KARACHI          HSSC     BS   
4320             Ho. No. 996 St. No. 67 Sultana abad Khr       A level     BS   
4933   House No. 694 St. No. 13 Cement factory colone...         F.Sc.     BS   
5623          House No.285 St. No.23  Dhamiyal Chishtian         F.Sc.     BS   
5852           Ho. No. 168 St. No. 97 Garrizon Sadiqabad       A level     BS   
7117                 House No.483 St. No.37  Dehri Thata            BS     MS   
7194                 h no. 292 St. No. 38 Koral Umer

Most of the values have 29 Feb as a date in an year which wasn't a **Leap Year**. We can correct them by decrementing the day.

In [772]:
sum(dob.isnull())

16

## No. of Invalid Dates: 16

## Total No. of Courses: 120

In [129]:
df_reg.Course.nunique()

120

## Total no. of Male vs Female students

In [807]:
num_male = sum(df_std.Gender == 'M')
num_female = sum(df_std.Gender == 'F')

In [808]:
pd.DataFrame.from_dict({"Males": num_male, "Females": num_female}, orient="index", columns=['Count']).T

Males  Females
Count   8935     4463

This doesn't contain the data from Peshawar Campus.

## Total No. of people who has taken more than 5 courses in a semester: 41200

In [89]:
grouped_semester_course = df_reg.groupby(['SID', 'Semester', 'Year']).agg({'Course': ['count']})
grouped_semester_course.columns = ['count']
grouped_semester_course = grouped_semester_course.reset_index()

In [90]:
sum(grouped_semester_course['count'] > 5)

41200

## The relationship between total no. of unique student ID’s and total no. of students



In [810]:
total_std = df_std.SID.count()
unq_ids = df_std.SID.str.extract('(\d+)').nunique()

In [811]:

pd.DataFrame.from_dict({"Unique IDs": unq_ids[0], "Total Students": total_std}, orient="index", columns=['Count']).T

Unique IDs  Total Students
Count        6601           17102

## Average no of people per semester of each campus

In [813]:
grouped_semester_students = df_reg.groupby(['Campus', 'Semester', 'Year']).agg({'SID': ['count']})
grouped_semester_students.columns = ['count']
grouped_semester_students = grouped_semester_students.reset_index()
grouped_semester_students

Campus Semester                        Year  count
0     KHR     Fall  1994-01-01 00:00:00.000000   3600
1     KHR     Fall  1995-01-01 00:00:00.000000   7200
2     KHR     Fall  1996-01-01 00:00:00.000000   7200
3     KHR     Fall  1997-01-01 00:00:00.000000   7200
4     KHR     Fall  1998-01-01 00:00:00.000000   3600
5     KHR     Fall  2001-01-01 00:00:00.000000   1200
6     KHR     Fall  2002-01-01 00:00:00.000000   2400
7     KHR     Fall  2003-01-01 00:00:00.000000   2400
8     KHR     Fall  2004-01-01 00:00:00.000000   2400
9     KHR   Spring  1995-01-01 00:00:00.000000   3600
10    KHR   Spring  1996-01-01 00:00:00.000000   7200
11    KHR   Spring  1997-01-01 00:00:00.000000   7200
12    KHR   Spring  1998-01-01 00:00:00.000000   7200
13    KHR   Spring  1999-01-01 00:00:00.000000   3600
14    KHR   Spring  2002-01-01 00:00:00.000000   1200
15    KHR   Spring  2003-01-01 00:00:00.000000   2400
16    KHR   Spring  2004-01-01 00:00:00.000000   2400
17    KHR   Spring  2005-01-01 00:00:00.000000   2400
18    LHR     Fall  1994-01-01 00:00:00.000000   2400
19    LHR     Fall  1995-01-01 00:00:00.000000   4800
20    LHR     Fall  1996-01-01 00:00:00.000000   7200
21    LHR     Fall  1997-01-01 00:00:00.000000   9600
22    LHR     Fall  1998-01-01 00:00:00.000000   9600
23    LHR     Fall  1999-01-01 00:00:00.000000   9600
24    LHR     Fall  2000-01-01 00:00:00.000000   9600
25    LHR     Fall  2001-01-01 00:00:00.000000   9600
26    LHR     Fall  2002-01-01 00:00:00.000000   9600
27    LHR     Fall  2003-01-01 00:00:00.000000   9600
28    LHR     Fall  2004-01-01 00:00:00.000000   9600
29    LHR     Fall  2020-01-01 00:00:00.000000   4200
30    LHR   Spring  1995-01-01 00:00:00.000000   2400
31    LHR   Spring  1996-01-01 00:00:00.000000   4800
32    LHR   Spring  1997-01-01 00:00:00.000000   7200
33    LHR   Spring  1998-01-01 00:00:00.000000   9600
34    LHR   Spring  1999-01-01 00:00:00.000000   9600
35    LHR   Spring  2000-01-01 00:00:00.000000   9600
36    LHR   Spring  2001-01-01 00:00:00.000000   9600
37    LHR   Spring  2002-01-01 00:00:00.000000   9600
38    LHR   Spring  2003-01-01 00:00:00.000000   9600
39    LHR   Spring  2004-01-01 00:00:00.000000   9600
40    LHR   Spring  2005-01-01 00:00:00.000000   9600
41    LHR   Spring  2020-01-01 00:00:00.000000   4200
42    PEW     Fall  1970-01-01 00:00:00.000001  32400
43    PEW     Fall  1970-01-01 00:00:00.000002  38100
44    PEW   Spring  1970-01-01 00:00:00.000000   7200
45    PEW   Spring  1970-01-01 00:00:00.000001  25200
46    PEW   Spring  1970-01-01 00:00:00.000002  38100

In [824]:
avg_semester_students = grouped_semester_students.reset_index().groupby(['Campus']).agg({'count': ['mean']})
avg_semester_students.columns = ['Average# Students']
# avg_semester_students = avg_semester_students.reset_index()
avg_semester_students

Average# Students
Campus                   
KHR           4133.333333
LHR           7950.000000
PEW          28200.000000

# Data Cleansing

- Separate first and last names both for the student and father. Standardize the first and last names.

In [694]:
names_list = df_std.Name.str.split().str
df_std["Std_First_Name"] = names_list[0:-1].str.join(' ')
df_std["Std_Last_Name"] = names_list[-1]

Let's look at the data where first name is empty.

In [695]:
df_std[df_std.Std_First_Name == ""]

SID       Name                       Father  \
1037   KHR_BS_1931        Ali                     Ali Ejaz   
1675   KHR_BS_2505     Namani                  Umar Namani   
2435    KHR_BS_319      Niazi                 Zubair Niazi   
2605   KHR_BS_3342   Hussaini        Ismail Munir Hussaini   
4257    KHR_BS_483      Munir           Munir Abdul Mujeeb   
4264   KHR_BS_4836    Jaffery      Aftab Mustansar Jaffery   
5755   KHR_BS_6178    Muazzam                Muazzam Tahir   
6170   KHR_BS_6551        Haq          Mouhammed Wajid Haq   
6879   KHR_MS_1248   Hussaini        Ismail Munir Hussaini   
7445    KHR_MS_318     Namani                  Umar Namani   
8833   LHR_BS_1567    Qureshi                Athar Qureshi   
10435  LHR_BS_3008      Haque  Nawabzada Adul Qudoos Haque   
12011   LHR_BS_467    Kiyyani              Tanveer Kiyyani   
13962  PEW_BS_1501     Ansari            Muaz Affan Ansari   
14919  PEW_BS_2363       Awan            Jameel Osman Awan   
15445  PEW_BS_2837      Waqas                  Waqas Jabir   
15491  PEW_BS_2879      Haque           Momd. Husain Haque   
15700  PEW_BS_3066     Ansary                 Sonil Ansary   
15911  PEW_BS_3256      Shabi                  Shabi Babar   
16219   PEW_BS_564    Kiyyani        Mohammed Ajab Kiyyani   
16635   PEW_BS_939     Abuzar        Abuzar Shams-ul-Islam   
16879   PEW_MS_258   Khanzada         Zafar Amjad Khanzada   

                              DoB Gender                    Reg Date  \
1037          1977-01-12 00:00:00      F  1997-08-23 00:00:00.000000   
1675          1979-10-11 00:00:00      F  1998-08-12 00:00:00.000000   
2435          1976-06-08 00:00:00      F  1994-08-08 00:00:00.000000   
2605          1979-10-23 00:00:00      F  1999-08-20 00:00:00.000000   
4257          1975-04-11 00:00:00      F  1994-08-21 00:00:00.000000   
4264          1982-09-24 00:00:00      F  2002-08-26 00:00:00.000000   
5755          1984-07-05 00:00:00      F  2004-08-01 00:00:00.000000   
6170          1984-10-13 00:00:00      F  2004-08-01 00:00:00.000000   
6879          1981-11-21 00:00:00      F  2004-08-26 00:00:00.000000   
7445          1979-09-07 00:00:00      F  2001-08-05 00:00:00.000000   
8833   1978-05-16 00:00:00.000000      F  1997-08-23 00:00:00.000000   
10435  1983-02-08 00:00:00.000000      F  2001-08-02 00:00:00.000000   
12011  1977-06-21 00:00:00.000000      F  1995-08-27 00:00:00.000000   
13962  1979-10-19 00:00:00.000000   None  1999-11-08 00:00:00.000000   
14919  1983-05-12 00:00:00.000000   None  2001-08-13 00:00:00.000000   
15445  1985-09-24 00:00:00.000000   None  2003-08-24 00:00:00.000000   
15491  1983-10-19 00:00:00.000000   None  2003-08-16 00:00:00.000000   
15700  1986-02-24 00:00:00.000000   None  2004-08-17 00:00:00.000000   
15911  1986-01-03 00:00:00.000000   None  2004-08-29 00:00:00.000000   
16219  1975-01-28 00:00:00.000000   None  1995-08-25 00:00:00.000000   
16635  1977-07-10 00:00:00.000000   None  1997-02-08 00:00:00.000000   
16879  1979-01-18 00:00:00.000000   None  2003-12-08 00:00:00.000000   

      Reg Status Degree Status  \
1037           A             C   
1675           A             C   
2435           A             C   
2605           A             C   
4257           A             C   
4264           A             I   
5755           A             I   
6170           A             I   
6879           A             I   
7445           A             C   
8833           A             C   
10435          A             C   
12011          A             C   
13962          A             C   
14919          A             C   
15445          A             I   
15491          A             I   
15700          A             I   
15911          A             I   
16219          A             C   
16635          A             C   
16879          A             C   

                                                 Address Qualification Degree  \
1037        house # 354 Street No.71  Odean chowk Multan           

By looking at the data, it seems that Father's name is typed in place of Student's Name by mistake. We can interchange their order, and then try to extract first and last names again.

In [696]:
m = df_std.Std_First_Name == ''
df_std.loc[m, ['Name', 'Father']] = (df.loc[m, ['Father', 'Name']].values)
df_std[df_std.Std_First_Name == ""]

SID                         Name     Father  \
1037   KHR_BS_1931                     Ali Ejaz        Ali   
1675   KHR_BS_2505                  Umar Namani     Namani   
2435    KHR_BS_319                 Zubair Niazi      Niazi   
2605   KHR_BS_3342        Ismail Munir Hussaini   Hussaini   
4257    KHR_BS_483           Munir Abdul Mujeeb      Munir   
4264   KHR_BS_4836      Aftab Mustansar Jaffery    Jaffery   
5755   KHR_BS_6178                Muazzam Tahir    Muazzam   
6170   KHR_BS_6551          Mouhammed Wajid Haq        Haq   
6879   KHR_MS_1248        Ismail Munir Hussaini   Hussaini   
7445    KHR_MS_318                  Umar Namani     Namani   
8833   LHR_BS_1567                Athar Qureshi    Qureshi   
10435  LHR_BS_3008  Nawabzada Adul Qudoos Haque      Haque   
12011   LHR_BS_467              Tanveer Kiyyani    Kiyyani   
13962  PEW_BS_1501            Muaz Affan Ansari     Ansari   
14919  PEW_BS_2363            Jameel Osman Awan       Awan   
15445  PEW_BS_2837                  Waqas Jabir      Waqas   
15491  PEW_BS_2879           Momd. Husain Haque      Haque   
15700  PEW_BS_3066                 Sonil Ansary     Ansary   
15911  PEW_BS_3256                  Shabi Babar      Shabi   
16219   PEW_BS_564        Mohammed Ajab Kiyyani    Kiyyani   
16635   PEW_BS_939        Abuzar Shams-ul-Islam     Abuzar   
16879   PEW_MS_258         Zafar Amjad Khanzada   Khanzada   

                              DoB Gender                    Reg Date  \
1037          1977-01-12 00:00:00      F  1997-08-23 00:00:00.000000   
1675          1979-10-11 00:00:00      F  1998-08-12 00:00:00.000000   
2435          1976-06-08 00:00:00      F  1994-08-08 00:00:00.000000   
2605          1979-10-23 00:00:00      F  1999-08-20 00:00:00.000000   
4257          1975-04-11 00:00:00      F  1994-08-21 00:00:00.000000   
4264          1982-09-24 00:00:00      F  2002-08-26 00:00:00.000000   
5755          1984-07-05 00:00:00      F  2004-08-01 00:00:00.000000   
6170          1984-10-13 00:00:00      F  2004-08-01 00:00:00.000000   
6879          1981-11-21 00:00:00      F  2004-08-26 00:00:00.000000   
7445          1979-09-07 00:00:00      F  2001-08-05 00:00:00.000000   
8833   1978-05-16 00:00:00.000000      F  1997-08-23 00:00:00.000000   
10435  1983-02-08 00:00:00.000000      F  2001-08-02 00:00:00.000000   
12011  1977-06-21 00:00:00.000000      F  1995-08-27 00:00:00.000000   
13962  1979-10-19 00:00:00.000000   None  1999-11-08 00:00:00.000000   
14919  1983-05-12 00:00:00.000000   None  2001-08-13 00:00:00.000000   
15445  1985-09-24 00:00:00.000000   None  2003-08-24 00:00:00.000000   
15491  1983-10-19 00:00:00.000000   None  2003-08-16 00:00:00.000000   
15700  1986-02-24 00:00:00.000000   None  2004-08-17 00:00:00.000000   
15911  1986-01-03 00:00:00.000000   None  2004-08-29 00:00:00.000000   
16219  1975-01-28 00:00:00.000000   None  1995-08-25 00:00:00.000000   
16635  1977-07-10 00:00:00.000000   None  1997-02-08 00:00:00.000000   
16879  1979-01-18 00:00:00.000000   None  2003-12-08 00:00:00.000000   

      Reg Status Degree Status  \
1037           A             C   
1675           A             C   
2435           A             C   
2605           A             C   
4257           A             C   
4264           A             I   
5755           A             I   
6170           A             I   
6879           A             I   
7445           A             C   
8833           A             C   
10435          A             C   
12011          A             C   
13962          A             C   
14919          A             C   
15445          A             I   
15491          A             I   
15700          A             I   
15911          A             I   
16219          A             C   
16635          A             C   
16879          A             C   

                                                 Address Qualification Degree  \
1037        house # 354 Street No.71  Odean chowk Multan           

In [697]:
names_list = df_std.Name.str.split().str
df_std["Std_First_Name"] = names_list[0:-1].str.join(' ')
df_std["Std_Last_Name"] = names_list[-1]

In [698]:
df_std[df_std.Std_First_Name == ""]

Empty DataFrame
Columns: [SID, Name, Father, DoB, Gender, Reg Date, Reg Status, Degree Status, Address, Qualification, Degree, Campus, Std_First_Name, Std_Last_Name]
Index: []

### Father's First and Last Name

In [699]:
fnames_list = df_std.Father.str.split().str
df_std["Father_First_Name"] = fnames_list[0:-1].str.join(' ')
df_std["Father_Last_Name"] = fnames_list[-1]

If a Father only has `Last Name`, make it his `First Name`.

In [700]:
df_std[df_std.Father_First_Name == '']

SID                         Name     Father  \
1037   KHR_BS_1931                     Ali Ejaz        Ali   
1675   KHR_BS_2505                  Umar Namani     Namani   
2435    KHR_BS_319                 Zubair Niazi      Niazi   
2605   KHR_BS_3342        Ismail Munir Hussaini   Hussaini   
4257    KHR_BS_483           Munir Abdul Mujeeb      Munir   
4264   KHR_BS_4836      Aftab Mustansar Jaffery    Jaffery   
5755   KHR_BS_6178                Muazzam Tahir    Muazzam   
6170   KHR_BS_6551          Mouhammed Wajid Haq        Haq   
6879   KHR_MS_1248        Ismail Munir Hussaini   Hussaini   
7445    KHR_MS_318                  Umar Namani     Namani   
8833   LHR_BS_1567                Athar Qureshi    Qureshi   
10435  LHR_BS_3008  Nawabzada Adul Qudoos Haque      Haque   
12011   LHR_BS_467              Tanveer Kiyyani    Kiyyani   
13962  PEW_BS_1501            Muaz Affan Ansari     Ansari   
14919  PEW_BS_2363            Jameel Osman Awan       Awan   
15445  PEW_BS_2837                  Waqas Jabir      Waqas   
15491  PEW_BS_2879           Momd. Husain Haque      Haque   
15700  PEW_BS_3066                 Sonil Ansary     Ansary   
15911  PEW_BS_3256                  Shabi Babar      Shabi   
16219   PEW_BS_564        Mohammed Ajab Kiyyani    Kiyyani   
16635   PEW_BS_939        Abuzar Shams-ul-Islam     Abuzar   
16879   PEW_MS_258         Zafar Amjad Khanzada   Khanzada   

                              DoB Gender                    Reg Date  \
1037          1977-01-12 00:00:00      F  1997-08-23 00:00:00.000000   
1675          1979-10-11 00:00:00      F  1998-08-12 00:00:00.000000   
2435          1976-06-08 00:00:00      F  1994-08-08 00:00:00.000000   
2605          1979-10-23 00:00:00      F  1999-08-20 00:00:00.000000   
4257          1975-04-11 00:00:00      F  1994-08-21 00:00:00.000000   
4264          1982-09-24 00:00:00      F  2002-08-26 00:00:00.000000   
5755          1984-07-05 00:00:00      F  2004-08-01 00:00:00.000000   
6170          1984-10-13 00:00:00      F  2004-08-01 00:00:00.000000   
6879          1981-11-21 00:00:00      F  2004-08-26 00:00:00.000000   
7445          1979-09-07 00:00:00      F  2001-08-05 00:00:00.000000   
8833   1978-05-16 00:00:00.000000      F  1997-08-23 00:00:00.000000   
10435  1983-02-08 00:00:00.000000      F  2001-08-02 00:00:00.000000   
12011  1977-06-21 00:00:00.000000      F  1995-08-27 00:00:00.000000   
13962  1979-10-19 00:00:00.000000   None  1999-11-08 00:00:00.000000   
14919  1983-05-12 00:00:00.000000   None  2001-08-13 00:00:00.000000   
15445  1985-09-24 00:00:00.000000   None  2003-08-24 00:00:00.000000   
15491  1983-10-19 00:00:00.000000   None  2003-08-16 00:00:00.000000   
15700  1986-02-24 00:00:00.000000   None  2004-08-17 00:00:00.000000   
15911  1986-01-03 00:00:00.000000   None  2004-08-29 00:00:00.000000   
16219  1975-01-28 00:00:00.000000   None  1995-08-25 00:00:00.000000   
16635  1977-07-10 00:00:00.000000   None  1997-02-08 00:00:00.000000   
16879  1979-01-18 00:00:00.000000   None  2003-12-08 00:00:00.000000   

      Reg Status Degree Status  \
1037           A             C   
1675           A             C   
2435           A             C   
2605           A             C   
4257           A             C   
4264           A             I   
5755           A             I   
6170           A             I   
6879           A             I   
7445           A             C   
8833           A             C   
10435          A             C   
12011          A             C   
13962          A             C   
14919          A             C   
15445          A             I   
15491          A             I   
15700          A             I   
15911          A             I   
16219          A             C   
16635          A             C   
16879          A             C   

                                                 Address Qualification Degree  \
1037        house # 354 Street No.71  Odean chowk Multan           

In [701]:
m = df_std.Father_First_Name == ''
df_std.loc[m, ['Father_First_Name', 'Father_Last_Name']] = (df_std.loc[m, ['Father_Last_Name', 'Father_First_Name']].values)
df_std[df_std.Father_First_Name == ""]

Empty DataFrame
Columns: [SID, Name, Father, DoB, Gender, Reg Date, Reg Status, Degree Status, Address, Qualification, Degree, Campus, Std_First_Name, Std_Last_Name, Father_First_Name, Father_Last_Name]
Index: []

### Name Standardization

In [702]:
all_names = pd.concat([df_std.Std_First_Name, df_std.Std_Last_Name, df_std.Father_First_Name, df_std.Father_Last_Name], ignore_index=True)

In [703]:
len(all_names.unique())

12475

In [709]:
std_name_map = {}

std_name_map.update({"AAdil": 'Aadil'})
std_name_map.update({"AAmer": 'Aamer'})
std_name_map.update({"AAmir": 'Aamir'})
std_name_map.update({"AAqil": 'Aaqil'})
std_name_map.update({"AAmina": 'Aamina'})
std_name_map.update({"Ch\.": 'Choudhary', "Chodhary": "Choudhary"})
std_name_map.update({"M\.": 'Mohammad', "Mohammed": "Mohammad", 
                     "Momd\.": "Mohammad", "Mouhammad": "Mohammad", 
                     "Mouhammed": "Mohammad", "Muhammad": "Mohammad"})
std_name_map.update({"Sh\.": "Sheikh", "Sheik": 'Sheikh', 'Shk\.': 'Sheikh'})
std_name_map.update({"Syed": "Syyed"})

std_name_map

{'AAdil': 'Aadil',
 'AAmer': 'Aamer',
 'AAmir': 'Aamir',
 'AAqil': 'Aaqil',
 'AAmina': 'Aamina',
 'Ch\\.': 'Choudhary',
 'Chodhary': 'Choudhary',
 'M\\.': 'Mohammad',
 'Mohammed': 'Mohammad',
 'Momd\\.': 'Mohammad',
 'Mouhammad': 'Mohammad',
 'Mouhammed': 'Mohammad',
 'Muhammad': 'Mohammad',
 'Sh\\.': 'Sheikh',
 'Sheik': 'Sheikh',
 'Shk\\.': 'Sheikh',
 'Syed': 'Syyed'}

In [710]:
df_std.Std_First_Name = df_std.Std_First_Name.replace(std_name_map, regex=True)
# sorted(df_std.Std_First_Name.unique())

In [711]:
df_std.Std_Last_Name = df_std.Std_Last_Name.replace(std_name_map, regex=True)
df_std.Name = df_std.Name.replace(std_name_map, regex=True)
df_std.Father = df_std.Father.replace(std_name_map, regex=True)
df_std.Father_First_Name = df_std.Father_First_Name.replace(std_name_map, regex=True)
df_std.Father_Last_Name = df_std.Father_Last_Name.replace(std_name_map, regex=True)

---

- Since the gender information is missing for the Peshawar campus, you can use the student names to figure out the gender. 

### Our Approach
We created a simple LSTM DL model which was trained on Student's first names from Karachi and Lahore campus, and it predicted whether a student was 'Male' or 'Female' based on the name.

First we'll create a training dataset for the model from the Lahore and Karachi campuses, and then use the Peshawar campus for prediction.

In [712]:
csv_data = df_std[['Std_First_Name', 'Gender']][~df_std.Gender.isnull()]
csv_data

Std_First_Name Gender
0               Hussain      M
1          Sheikh Munir      M
2                 Hamna      F
3                Jabbar      M
4           Rana Hayyat      M
...                 ...    ...
13397              Rohi      F
13398      Mahtab Emmad      M
13399    Mohammad Latif      M
13400              Iman      F
13401  Choudhary Muneer      M

[13398 rows x 2 columns]

In [713]:
csv_data.to_csv(data_dir / 'names.csv', index=False)

In [714]:
null_gender = df_std[['Std_First_Name']][df_std.Gender.isnull()]
null_gender

Std_First_Name
11982                 Karam
12397            Abdul Ahad
12972                 Mehak
12991  Mohammad Abdur Rahim
13402          Midhat Azhar
...                     ...
17097               Natalia
17098     Agha Abdul kareem
17099                 Azhar
17100    Riffat Ajmal Affan
17101              Zulfiqar

[3704 rows x 1 columns]

In [715]:
pd.Series(null_gender.Std_First_Name.unique()).to_csv(data_dir / 'names_unlabeled.csv', index=False, header=False)

In [716]:
sum(null_gender.Std_First_Name == "")

0

### Load predicted genders

In [724]:
import json

with open(data_dir / 'genders.json') as json_file:
    pew_gender_map = json.load(json_file)

In [725]:
gender_values = null_gender.Std_First_Name.map(pew_gender_map)

In [727]:
df_std.loc[df_std.Gender.isnull(), 'Gender'] = gender_values

In [728]:
df_std[df_std.Campus == 'PEW']

SID                        Name                Father  \
13402     PEW_BS_0        Midhat Azhar Durrani         Azhar Durrani   
13403     PEW_BS_1  Mohammad Abdul Rafay Kazmi           Iffan Kazmi   
13404    PEW_BS_10                Ghosia Najum           Najum Yosaf   
13405   PEW_BS_100      Shuja Taimor Kharasani       Sahir Kharasani   
13406  PEW_BS_1000                Kashif Zaidi           Zaran Zaidi   
...            ...                         ...                   ...   
17097    PEW_MS_95               Natalia Baqir     Baqir Abdul Rafay   
17098    PEW_MS_96    Agha Abdul kareem Abbasi          Ghazi Abbasi   
17099    PEW_MS_97               Azhar Durrani      Ibraheem Durrani   
17100    PEW_MS_98   Riffat Ajmal Affan Ansary    Ajmal Affan Ansary   
17101    PEW_MS_99             Zulfiqar Gondal  Mohammad Ijaz Gondal   

                              DoB Gender                    Reg Date  \
13402  1974-04-14 00:00:00.000000      F  1994-08-26 00:00:00.000000   
13403  1974-12-13 00:00:00.000000      M  1994-08-13 00:00:00.000000   
13404  1974-04-25 00:00:00.000000      F  1994-08-14 00:00:00.000000   
13405  1974-05-22 00:00:00.000000      M  1994-08-18 00:00:00.000000   
13406  1978-02-25 00:00:00.000000      M  1997-08-28 00:00:00.000000   
...                           ...    ...                         ...   
17097  1979-03-15 00:00:00.000000      F  2001-01-08 00:00:00.000000   
17098  1978-11-15 00:00:00.000000      M  2001-05-08 00:00:00.000000   
17099  1978-11-17 00:00:00.000000      M  2001-08-08 00:00:00.000000   
17100  1978-08-20 00:00:00.000000      F  2001-11-08 00:00:00.000000   
17101  1978-04-20 00:00:00.000000      M  2001-08-14 00:00:00.000000   

      Reg Status Degree Status  \
13402          A             C   
13403          A             C   
13404          A             C   
13405          A             C   
13406          A             C   
...          ...           ...   
17097          A             C   
17098          A             C   
17099          A             C   
17100          A             C   
17101          A             C   

                                               Address Qualification Degree  \
13402             H# 447 St. # 5  Air port coloney PSH       A-Level     BS   
13403             H#  169 s#  0 University coloney PSH       A-Level     BS   
13404           h#  697 St. #  94 Lohi pura Hasanabdal       A-Level     BS   
13405                   H# 504 S No.4  Bara market PSH       A-Level     BS   
13406  H No.509 St. No.38  University coloney PESHAWAR          HSSC     BS   
...                                                ...           ...    ...   
17097     House No. 699 Street #  59 Chand mari Gilgit            BS     MS   
17098           H No. 14 s no. 55 Bara market PESHAWAR            MS     MS   
17099             Ho. no. 329 s no. 50 Khybar road Psh          M.Sc     MS   
17100          h no.875 St. no.45  Purani Taal Noshera            MS     MS   
17101         Ho. No.190 St. # 9  Lane 6 Mall road Psh          M.Sc     MS   

      Campus        Std_First_Name Std_Last_Name Father_First_Name  \
13402    PEW          Midhat Azhar       Durrani             Azhar   
13403    PEW  Mohammad Abdul Rafay         Kazmi             Iffan   
13404    PEW                Ghosia         Najum             Najum   
13405    PEW          Shuja Taimor     Kharasani             Sahir   
13406    PEW                Kashif         Zaidi             Zaran   
...      ...                   ...           ...               ...   
17097    PEW               Natalia         Baqir       Baqir Abdul   
17098    PEW     Agha Abdul kareem        Abbasi             Ghazi   
17099    PEW                 Azhar       Durrani          Ibraheem   
17100    PEW    Riffat Ajmal Affan        Ansary       Ajmal Affan   
17101    PEW              Zulfiqar        Gondal     Mohammad Ijaz   

      Father_Last_Name  
13402          Durrani  
13403            Kazmi  
13404      

In [729]:
sum(df_std.Gender.isnull())

0

### Each student has been assigned a gender.

---

- Introduce a new column i.e. city_name. It will involve extracting the city from the address and then the standardization of the city names. (Use the city names in the telephone directory as standard)

In [730]:
city_names = df_std.Address.str.split().str[-1]

In [731]:
last_two_city_names = df_std.Address.str.split().str[-2:].str.join(' ')
# sorted(last_two_city_names.unique())

In [732]:
sorted(city_names.unique())

['Abbotabad',
 'Attock',
 'Bahawalpur',
 'Bhatian',
 'Bhawalpur',
 'Chishtian',
 'Faisalabad',
 'Gilgit',
 'Gujrat',
 'Gujrawala',
 'Hakim',
 'Haripur',
 'Hasanabdal',
 'Hyderabad',
 'Jehlum',
 'KARACHI',
 'KHR',
 'Kabirwala',
 'Karachi',
 'Khan',
 'Kharian',
 'Khas',
 'Khr',
 'Kot',
 'LalaMusa',
 'Lyyah',
 'Mansehra',
 'Mardan',
 'Melsi',
 'Multan',
 'Noshera',
 'PESHAWAR',
 'PSH',
 'Peshawar',
 'Psh',
 'Quetta',
 'Sadiqabad',
 'Sahiwal',
 'Sargodha',
 'Sawat',
 'Shekhopura',
 'Sialkot',
 'Sibbi',
 'Thata',
 'Wahari',
 'Wazirabad',
 'karachi',
 'khan',
 'khr',
 'peshawar',
 'psh']

In [733]:
std_city_map = {}
std_city_map.update(
    dict.fromkeys(
        [
            'KHR', 
            'KARACHI', 
            'Khr', 
            'Karachi', 
            'khr', 
            'karachi'
            ], 'Karachi'))

std_city_map.update(
    dict.fromkeys(
        [
            'PSH',
            'Psh',
            'Peshawar',
            'PESHAWAR',
            'peshawar',
            'psh'
        ], 'Peshawar'
    )
)

std_city_map.update({'Abbotabad': 'Abottabad'})
std_city_map.update({'Bahawalpur': 'Bahawalpur', 'Bhawalpur': 'Bahawalpur'})
std_city_map.update({'Bhatian': 'Pindi Bhattian'})
std_city_map.update({'D.G. Khan': 'Dera Ghazi Khan', 'ghazi khan': 'Dera Ghazi Khan'})
std_city_map.update({'D.I. Khan': 'D.I.Khan', 'Ismail Khan': 'D.I.Khan'})
std_city_map.update({'Gujrawala': 'Gujranwala'})
std_city_map.update({'Hakim': 'Abdul Hakim'})
std_city_map.update({'Hasanabdal': 'Hasan Abdal'})
std_city_map.update({'Jehlum': 'Jhelum'})
std_city_map.update({'Khas': 'Mirpur Khas'})
std_city_map.update({'Lyyah': 'Layyah'})
std_city_map.update({'Melsi': 'Mailsi'})
std_city_map.update({'Noshera': 'Nowshera'})
std_city_map.update({'Sawat': 'Swat'})
std_city_map.update({'Shekhopura': 'Sheikhupura'})
std_city_map.update({'Sibbi': 'Sibi'})
std_city_map.update({'Thata': 'Thatta'})
std_city_map.update({'Wahari': 'Vehari'})

In [735]:
two_words_city_name = ['Khan', 'Kot']

In [736]:
# Get the previous word for these, in order to figure out the city
city_names[city_names == 'Khan'] = last_two_city_names[city_names == 'Khan']
city_names[city_names == 'khan'] = last_two_city_names[city_names == 'khan']
city_names[city_names == 'Kot'] = last_two_city_names[city_names == 'Kot']

In [738]:
std_city_names = city_names.replace(std_city_map)

In [739]:
sorted(std_city_names.unique())

['Abdul Hakim',
 'Abottabad',
 'Attock',
 'Bahawalpur',
 'Chishtian',
 'D.I.Khan',
 'Dera Ghazi Khan',
 'Faisalabad',
 'Gilgit',
 'Gujranwala',
 'Gujrat',
 'Haripur',
 'Hasan Abdal',
 'Hyderabad',
 'Jhelum',
 'Kabirwala',
 'Karachi',
 'Kharian',
 'LalaMusa',
 'Layyah',
 'Mailsi',
 'Mansehra',
 'Mardan',
 'Mirpur Khas',
 'Multan',
 'Nowshera',
 'Peshawar',
 'Pindi Bhattian',
 'Quetta',
 'Sadiqabad',
 'Sahiwal',
 'Sargodha',
 'Sheikhupura',
 'Sialkot',
 'Sibi',
 'Swat',
 'Thatta',
 'Umer Kot',
 'Usman Kot',
 'Vehari',
 'Wazirabad']

In [740]:
df_std['City'] = std_city_names

In [741]:
df_std.head()

SID                   Name                Father  \
0     KHR_BS_0         Hussain Ansary          Ubaid Ansary   
1     KHR_BS_1  Sheikh Munir Hussaini  Viqar Hamid Hussaini   
2    KHR_BS_10           Hamna Ansary         Hameed Ansary   
3   KHR_BS_100          Jabbar Haqqie     Muneer Rai Haqqie   
4  KHR_BS_1000       Rana Hayyat Baig   Ghulam Mustafa Baig   

                   DoB Gender                    Reg Date Reg Status  \
0  1974-01-05 00:00:00      M  1994-08-13 00:00:00.000000          A   
1  1974-12-13 00:00:00      M  1994-08-13 00:00:00.000000          A   
2  1974-04-25 00:00:00      F  1994-08-14 00:00:00.000000          A   
3  1974-05-22 00:00:00      M  1994-08-18 00:00:00.000000          A   
4  1976-02-25 00:00:00      M  1995-08-28 00:00:00.000000          A   

  Degree Status                                            Address  \
0             C           h# 978 Street No.72  Defense Phase 1 KHR   
1             C                    H#  169 s#  0 Mutian wala Thata   
2             C                     h#  697 St. #  94 sea site KHR   
3             C                       H# 504 S No.4  Kumria Quetta   
4             C  H No.509 St. No.38  Kishwar heights university...   

  Qualification Degree Campus Std_First_Name Std_Last_Name Father_First_Name  \
0       A-Level     BS    KHR        Hussain        Ansary             Ubaid   
1       A-Level     BS    KHR   Sheikh Munir      Hussaini       Viqar Hamid   
2       A-Level     BS    KHR          Hamna        Ansary            Hameed   
3       A-Level     BS    KHR         Jabbar        Haqqie        Muneer Rai   
4          HSSC     BS    KHR    Rana Hayyat          Baig    Ghulam Mustafa   

  Father_Last_Name     City  
0           Ansary  Karachi  
1         Hussaini   Thatta  
2           Ansary  Karachi  
3           Haqqie   Quetta  
4             Baig  Karachi

In [742]:
df_std.describe()

SID           Name       Father                         DoB  \
count         17102          17102        17102                       17100   
unique        17102          14642        14230                        7380   
top     PEW_BS_1573  Kishwar Pasha  Amjad Satti  1981-03-22 00:00:00.000000   
freq              1              6            5                           8   

       Gender                    Reg Date Reg Status Degree Status  \
count   17102                       17102      17102         17102   
unique      2                         442          2             2   
top         M  2003-08-16 00:00:00.000000          A             C   
freq    11531                          76      17100         12500   

                                       Address Qualification Degree Campus  \
count                                    17102         17102  17102  17102   
unique                                   15360            12      2      3   
top     house no.141 S# 16  Haider road Multan         F.Sc.     BS    KHR   
freq                                         3          2390  14302   8201   

       Std_First_Name Std_Last_Name Father_First_Name Father_Last_Name  \
count           17102         17102             17102            17102   
unique           5879           315              6458              316   
top           Firdous         Zaidi             Amjad            Zaidi   
freq               55           369                62              369   

           City  
count     17102  
unique       41  
top     Karachi  
freq       4438

### We have 41 Unique cities

---

- Bring the gender information into a consistent representation. Use ‘M’ and ‘F’ with data type CHAR(1).


We've already achieved this during the data collection phase.

## Date Validation

- If you have been using VARCHAR or CHAR for date, port the date information into columns with the proper Date data type. The dates should be as per calendar dates.

In [743]:
df_std["Reg Date"] = pd.to_datetime(df_std["Reg Date"])

In [744]:
df_reg.Year = pd.to_datetime(df_reg.Year)

In [745]:
df_std.DoB = pd.to_datetime(df_std.DoB)

ParserError: day is out of range for month: 29/Feb/78

Converting `DoB` to `DateTime` gives error because of invalid dates, i.e. `27/Feb/Non-Leap-Year`. We'll fix this in the next section.

---

- Validate all the dates against the business rules e.g. the DOB should be smaller than the Reg. Date and Graduation Date should be greater than Reg. Date. The data might be having anomalies like exchanged DOB and Reg. Date by mistake etc. Also be careful with invalid dates i.e. 31st Feb. or 29th Feb. in a non-leap year.

First, let's extract invalid `DoB` dates.

In [746]:
dob = pd.to_datetime(df_std.DoB, errors="coerce")

In [747]:
df_std[dob.isnull()]

SID                          Name  \
516    KHR_BS_1462                   Annan Satti   
593    KHR_BS_1531                 Hadiqa Abuzar   
1282   KHR_BS_2151                    Affan Baig   
2113    KHR_BS_290                   Qasim Rehan   
3631   KHR_BS_4266                  Shafaq Ahmed   
4167   KHR_BS_4749              Sumiyya Khanzada   
4319   KHR_BS_4886           Khushbakht Khanzada   
4320   KHR_BS_4887               Shabir Hussaini   
4933   KHR_BS_5438       Aga Abdus Sammi Khattak   
5623   KHR_BS_6059                Arsalan Chohan   
5852   KHR_BS_6265                  Amina Muslim   
7117   KHR_MS_1462        Mohammad Ghani Paracha   
7194   KHR_MS_1531                    Arif Amjad   
7414    KHR_MS_290             Sonia Shaheen Haq   
10649  LHR_BS_3200  Choudhary Abdul Qadir Minhas   
11692   LHR_BS_414      Pervaiz Sikandar Sherazi   

                             Father        DoB Gender   Reg Date Reg Status  \
516               Husain Alam Satti  29/Feb/78      M 1996-08-28          A   
593                  Abuzar Zeeshan  29/Feb/77      F 1996-08-22          A   
1282        Mohammad Sheharyar Baig  29/Feb/79      M 1997-08-16          A   
2113                   Rehan Nabeel  29/Feb/75      M 1994-08-29          A   
3631                     Ahmed Ajab  29/Feb/82      F 2001-08-21          A   
4167                  Arif Khanzada  29/Feb/83      F 2001-08-03          A   
4319         Sheikh Sarwar Khanzada  29/Feb/83      F 2002-08-16          A   
4320         Syyed Muazzum Hussaini  29/Feb/83      M 2002-08-19          A   
4933     Abdul Qadeer Ikram Khattak  29/Feb/85      M 2003-08-20          A   
5623              Rabbi Daud Chohan  29/Feb/86      M 2004-08-18          A   
5852                 Muslim Shujaat  29/Feb/85      F 2004-08-24          A   
7117           Rehmat Kabir Paracha  29/Feb/82      M 2004-08-28          A   
7194                  Amjad Inayyat  29/Feb/81      M 2004-08-22          A   
7414                    Shaheen Haq  29/Feb/78      F 2001-08-29          A   
10649                  Viqar Minhas       None      M 2002-08-20          A   
11692  Abdul Mannan Shabeer Sherazi       None      M 1995-08-28          A   

      Degree Status                                            Address  \
516               C                House No.483 St. No.37  Dehri Thata   
593               C                h no. 292 St. No. 38 Koral Umer Kot   
1282              C       house no.568 St. No.52  Zeenat block Karachi   
2113              C      Ho. No. 193 St. No. 14 Chand mari Mirpur Khas   
3631              C          Ho. No.405 St. No.57  Doley shah Umer Kot   
4167              C   house no. 609 St. No. 98 Shah latif town Karachi   
4319              I            h no. 681 St. No. 4 Pwd Coloney KARACHI   
4320              I            Ho. No. 996 St. No. 67 Sultana abad Khr   
4933              I  House No. 694 St. No. 13 Cement factory colone...   
5623              I         House No.285 St. No.23  Dhamiyal Chishtian   
5852              I          Ho. No. 168 St. No. 97 Garrizon Sadiqabad   
7117              I                House No.483 St. No.37  Dehri Thata   
7194              I                h no. 292 St. No. 38 Koral Umer Kot   
7414              C                        Ho. No. 193 St. No. 14  Khr   
10649             I      House No.117 St. no.41  Malik abad Bahawalpur   
11692             C             h no.249 St. no.77  Ramzania Wazirabad   

      Qualification Degree Campus         Std_First_Name Std_Last_Name  \
516           F.Sc.     BS    KHR                  Annan         Satti   
593            HSSC     BS    KHR                 Hadiqa        Abuzar   
1282          F.Sc.     BS    KHR                  Affan          Baig   
2113        A level     BS    KHR                  Qasim         Rehan   
3631        A level     BS    KHR                 Shafaq         Ahmed   
4167          F.Sc.     BS    KHR                Sumiyya      Khanzada   
4319     

Except the last two who have missing dates, all invalid dates have `29 Feb` in a Non-Leap year. We can fix it by making the dates `28`.

In [748]:
df_std.loc[dob.isnull(), 'DoB'] = df_std.loc[dob.isnull(), 'DoB'].str.replace('29', '28')

In [749]:
df_std[dob.isnull()]

SID                          Name  \
516    KHR_BS_1462                   Annan Satti   
593    KHR_BS_1531                 Hadiqa Abuzar   
1282   KHR_BS_2151                    Affan Baig   
2113    KHR_BS_290                   Qasim Rehan   
3631   KHR_BS_4266                  Shafaq Ahmed   
4167   KHR_BS_4749              Sumiyya Khanzada   
4319   KHR_BS_4886           Khushbakht Khanzada   
4320   KHR_BS_4887               Shabir Hussaini   
4933   KHR_BS_5438       Aga Abdus Sammi Khattak   
5623   KHR_BS_6059                Arsalan Chohan   
5852   KHR_BS_6265                  Amina Muslim   
7117   KHR_MS_1462        Mohammad Ghani Paracha   
7194   KHR_MS_1531                    Arif Amjad   
7414    KHR_MS_290             Sonia Shaheen Haq   
10649  LHR_BS_3200  Choudhary Abdul Qadir Minhas   
11692   LHR_BS_414      Pervaiz Sikandar Sherazi   

                             Father        DoB Gender   Reg Date Reg Status  \
516               Husain Alam Satti  28/Feb/78      M 1996-08-28          A   
593                  Abuzar Zeeshan  28/Feb/77      F 1996-08-22          A   
1282        Mohammad Sheharyar Baig  28/Feb/79      M 1997-08-16          A   
2113                   Rehan Nabeel  28/Feb/75      M 1994-08-29          A   
3631                     Ahmed Ajab  28/Feb/82      F 2001-08-21          A   
4167                  Arif Khanzada  28/Feb/83      F 2001-08-03          A   
4319         Sheikh Sarwar Khanzada  28/Feb/83      F 2002-08-16          A   
4320         Syyed Muazzum Hussaini  28/Feb/83      M 2002-08-19          A   
4933     Abdul Qadeer Ikram Khattak  28/Feb/85      M 2003-08-20          A   
5623              Rabbi Daud Chohan  28/Feb/86      M 2004-08-18          A   
5852                 Muslim Shujaat  28/Feb/85      F 2004-08-24          A   
7117           Rehmat Kabir Paracha  28/Feb/82      M 2004-08-28          A   
7194                  Amjad Inayyat  28/Feb/81      M 2004-08-22          A   
7414                    Shaheen Haq  28/Feb/78      F 2001-08-29          A   
10649                  Viqar Minhas       None      M 2002-08-20          A   
11692  Abdul Mannan Shabeer Sherazi       None      M 1995-08-28          A   

      Degree Status                                            Address  \
516               C                House No.483 St. No.37  Dehri Thata   
593               C                h no. 292 St. No. 38 Koral Umer Kot   
1282              C       house no.568 St. No.52  Zeenat block Karachi   
2113              C      Ho. No. 193 St. No. 14 Chand mari Mirpur Khas   
3631              C          Ho. No.405 St. No.57  Doley shah Umer Kot   
4167              C   house no. 609 St. No. 98 Shah latif town Karachi   
4319              I            h no. 681 St. No. 4 Pwd Coloney KARACHI   
4320              I            Ho. No. 996 St. No. 67 Sultana abad Khr   
4933              I  House No. 694 St. No. 13 Cement factory colone...   
5623              I         House No.285 St. No.23  Dhamiyal Chishtian   
5852              I          Ho. No. 168 St. No. 97 Garrizon Sadiqabad   
7117              I                House No.483 St. No.37  Dehri Thata   
7194              I                h no. 292 St. No. 38 Koral Umer Kot   
7414              C                        Ho. No. 193 St. No. 14  Khr   
10649             I      House No.117 St. no.41  Malik abad Bahawalpur   
11692             C             h no.249 St. no.77  Ramzania Wazirabad   

      Qualification Degree Campus         Std_First_Name Std_Last_Name  \
516           F.Sc.     BS    KHR                  Annan         Satti   
593            HSSC     BS    KHR                 Hadiqa        Abuzar   
1282          F.Sc.     BS    KHR                  Affan          Baig   
2113        A level     BS    KHR                  Qasim         Rehan   
3631        A level     BS    KHR                 Shafaq         Ahmed   
4167          F.Sc.     BS    KHR                Sumiyya      Khanzada   
4319     

In [750]:
dob = pd.to_datetime(df_std.DoB, errors="coerce")
df_std[dob.isnull()]

SID                          Name  \
10649  LHR_BS_3200  Choudhary Abdul Qadir Minhas   
11692   LHR_BS_414      Pervaiz Sikandar Sherazi   

                             Father   DoB Gender   Reg Date Reg Status  \
10649                  Viqar Minhas  None      M 2002-08-20          A   
11692  Abdul Mannan Shabeer Sherazi  None      M 1995-08-28          A   

      Degree Status                                        Address  \
10649             I  House No.117 St. no.41  Malik abad Bahawalpur   
11692             C         h no.249 St. no.77  Ramzania Wazirabad   

      Qualification Degree Campus         Std_First_Name Std_Last_Name  \
10649         F.Sc.     BS    LHR  Choudhary Abdul Qadir        Minhas   
11692          HSSC     BS    LHR       Pervaiz Sikandar       Sherazi   

          Father_First_Name Father_Last_Name        City  
10649                 Viqar           Minhas  Bahawalpur  
11692  Abdul Mannan Shabeer          Sherazi   Wazirabad

To figure out the missing dates, we can deduce it from the `Reg Date`, and take the median DoB of all the students who registered around the same time.

In [751]:
def get_median_dob(reg_date):
    a = pd.Timestamp(reg_date) - pd.Timedelta(5, unit='d')
    b = pd.Timestamp(reg_date) + pd.Timedelta(5, unit='d')
    df2_numeric_format = pd.to_numeric(pd.to_datetime(df_std[df_std["Reg Date"].between(a, b)].DoB))
    median_numeric_format = np.median(df2_numeric_format)
    median_datetime_format = pd.to_datetime(median_numeric_format)
    return median_datetime_format

In [752]:
get_median_dob("1995-08-28")

Timestamp('1976-08-06 00:00:00')

In [753]:
invalid_dates = df_std.loc[dob.isnull()][['DoB', 'Reg Date']]
invalid_dates.DoB = invalid_dates['Reg Date'].apply(get_median_dob)
invalid_dates

DoB   Reg Date
10649 1982-07-25 12:00:00 2002-08-20
11692 1976-08-06 00:00:00 1995-08-28

In [754]:
df_std.loc[dob.isnull(), 'DoB'] = df_std.loc[dob.isnull(), 'Reg Date'].apply(get_median_dob)

In [755]:
dob = pd.to_datetime(df_std.DoB, errors="coerce")
df_std[dob.isnull()]

Empty DataFrame
Columns: [SID, Name, Father, DoB, Gender, Reg Date, Reg Status, Degree Status, Address, Qualification, Degree, Campus, Std_First_Name, Std_Last_Name, Father_First_Name, Father_Last_Name, City]
Index: []

Now that all dates are valid, we'll change the data type of column to `DateTime`.

In [756]:
df_std.DoB = pd.to_datetime(df_std.DoB)

In [757]:
df_std.head()

SID                   Name                Father        DoB Gender  \
0     KHR_BS_0         Hussain Ansary          Ubaid Ansary 1974-01-05      M   
1     KHR_BS_1  Sheikh Munir Hussaini  Viqar Hamid Hussaini 1974-12-13      M   
2    KHR_BS_10           Hamna Ansary         Hameed Ansary 1974-04-25      F   
3   KHR_BS_100          Jabbar Haqqie     Muneer Rai Haqqie 1974-05-22      M   
4  KHR_BS_1000       Rana Hayyat Baig   Ghulam Mustafa Baig 1976-02-25      M   

    Reg Date Reg Status Degree Status  \
0 1994-08-13          A             C   
1 1994-08-13          A             C   
2 1994-08-14          A             C   
3 1994-08-18          A             C   
4 1995-08-28          A             C   

                                             Address Qualification Degree  \
0           h# 978 Street No.72  Defense Phase 1 KHR       A-Level     BS   
1                    H#  169 s#  0 Mutian wala Thata       A-Level     BS   
2                     h#  697 St. #  94 sea site KHR       A-Level     BS   
3                       H# 504 S No.4  Kumria Quetta       A-Level     BS   
4  H No.509 St. No.38  Kishwar heights university...          HSSC     BS   

  Campus Std_First_Name Std_Last_Name Father_First_Name Father_Last_Name  \
0    KHR        Hussain        Ansary             Ubaid           Ansary   
1    KHR   Sheikh Munir      Hussaini       Viqar Hamid         Hussaini   
2    KHR          Hamna        Ansary            Hameed           Ansary   
3    KHR         Jabbar        Haqqie        Muneer Rai           Haqqie   
4    KHR    Rana Hayyat          Baig    Ghulam Mustafa             Baig   

      City  
0  Karachi  
1   Thatta  
2  Karachi  
3   Quetta  
4  Karachi

### DOB should be smaller than Reg Date

In [758]:
df_std[df_std.DoB > df_std['Reg Date']]

Empty DataFrame
Columns: [SID, Name, Father, DoB, Gender, Reg Date, Reg Status, Degree Status, Address, Qualification, Degree, Campus, Std_First_Name, Std_Last_Name, Father_First_Name, Father_Last_Name, City]
Index: []

---

- For some campuses, the degree information is missing. Devise some technique to figure it out and update the rows with empty degree fields.

In [759]:
sum(df_std.Degree.isnull())

0

No Degree information is missing.

---

- Validate other businessrules for each field. One example is that marks should be in the range 0 to 100 inclusive.

In [760]:
df_std.describe(include='all')

SID           Name       Father                  DoB Gender  \
count         17102          17102        17102                17102  17102   
unique        17102          14642        14230                 4085      2   
top     PEW_BS_1573  Kishwar Pasha  Amjad Satti  1982-03-28 00:00:00      M   
freq              1              6            5                   14  11531   
first           NaN            NaN          NaN  1974-01-01 00:00:00    NaN   
last            NaN            NaN          NaN  1986-12-29 00:00:00    NaN   

                   Reg Date Reg Status Degree Status  \
count                 17102      17102         17102   
unique                  442          2             2   
top     2003-08-16 00:00:00          A             C   
freq                     76      17100         12500   
first   1994-01-08 00:00:00        NaN           NaN   
last    2005-01-12 00:00:00        NaN           NaN   

                                       Address Qualification Degree Campus  \
count                                    17102         17102  17102  17102   
unique                                   15360            12      2      3   
top     house no.141 S# 16  Haider road Multan         F.Sc.     BS    KHR   
freq                                         3          2390  14302   8201   
first                                      NaN           NaN    NaN    NaN   
last                                       NaN           NaN    NaN    NaN   

       Std_First_Name Std_Last_Name Father_First_Name Father_Last_Name  \
count           17102         17102             17102            17102   
unique           5879           315              6458              316   
top           Firdous         Zaidi             Amjad            Zaidi   
freq               55           369                62              369   
first             NaN           NaN               NaN              NaN   
last              NaN           NaN               NaN              NaN   

           City  
count     17102  
unique       41  
top     Karachi  
freq       4438  
first       NaN  
last        NaN

In [761]:
df_reg.Score.describe(include='all')

count    406200.000000
mean         74.460268
std          14.415844
min          50.000000
25%          62.000000
50%          74.000000
75%          87.000000
max          99.000000
Name: Score, dtype: float64

# Save Cleaned Data in DB

In [762]:
from sqlalchemy import create_engine
from sqlalchemy.types import Integer, Text, String, DateTime, Boolean, CHAR

In [763]:
df_reg.columns

Index(['SID', 'Course', 'Score', 'Semester', 'Year', 'Discipline', 'Degree',
       'Campus'],
      dtype='object')

In [764]:
std_col_types = {
    "SID": Text,
    "Name": Text,
    "Father": Text,
    "DoB":  DateTime,
    "Gender": CHAR(1),
    "Reg Date": DateTime,
    "Reg Status": CHAR(1),
    "Degree Status": CHAR(1),
    "Address": Text,
    "Qualification": String(8),
    "Degree": CHAR(2),
    "Campus": CHAR(2),
    "Std_First_Name": Text,
    "Std_Last_Name": Text,
    "Father_First_Name": Text,
    "Father_Last_Name": Text
    }

In [765]:
reg_col_types = {
    "SID": Text,
    "Courses": String(8),
    "Score": Integer,
    "Semester":  String(10),
    "Year": DateTime,
    "Discipline": String(3),
    "Degree": CHAR(2),
    "Campus": CHAR(2)
    }

In [766]:
engine = create_engine('sqlite:///University.db', echo=True)
sqlite_connection = engine.connect()

std_table = "CLN_Students"
df_std.to_sql(
    std_table, 
    sqlite_connection, 
    if_exists='replace', 
    index=False, 
    dtype=std_col_types)


reg_table = "CLN_Course_Registrations"
df_reg.to_sql(
    reg_table, 
    sqlite_connection, 
    if_exists='replace', 
    index=False, 
    dtype=reg_col_types)


sqlite_connection.close()

2020-11-16 16:11:08,724 INFO sqlalchemy.engine.base.Engine SELECT CAST('test plain returns' AS VARCHAR(60)) AS anon_1
2020-11-16 16:11:08,725 INFO sqlalchemy.engine.base.Engine ()
2020-11-16 16:11:08,726 INFO sqlalchemy.engine.base.Engine SELECT CAST('test unicode returns' AS VARCHAR(60)) AS anon_1
2020-11-16 16:11:08,727 INFO sqlalchemy.engine.base.Engine ()
2020-11-16 16:11:08,734 INFO sqlalchemy.engine.base.Engine PRAGMA main.table_info("CLN_Students")
2020-11-16 16:11:08,735 INFO sqlalchemy.engine.base.Engine ()
2020-11-16 16:11:08,737 INFO sqlalchemy.engine.base.Engine PRAGMA main.table_info("CLN_Students")
2020-11-16 16:11:08,738 INFO sqlalchemy.engine.base.Engine ()
2020-11-16 16:11:08,740 INFO sqlalchemy.engine.base.Engine SELECT name FROM sqlite_master WHERE type='table' ORDER BY name
2020-11-16 16:11:08,741 INFO sqlalchemy.engine.base.Engine ()
2020-11-16 16:11:08,743 INFO sqlalchemy.engine.base.Engine PRAGMA main.table_xinfo("CLN_Students")
2020-11-16 16:11:08,744 INFO sqlal